content/en/docs/tasks/observability/metrics/querying-metrics/test.sh

```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env bash
+# shellcheck disable=SC1090,SC2154,SC2155
+
+# Copyright Istio Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e
+set -u
+set -o pipefail
+
+source "tests/util/samples.sh"
+
+# @setup profile=demo
+kubectl label namespace default istio-injection=enabled --overwrite
+
+# Install Bookinfo sample
+startup_bookinfo_sample
+
+# Install Prometheus
+kubectl apply -f samples/addons/prometheus.yaml -n istio-system
+_wait_for_deployment istio-system prometheus
+
+# Fire a couple of requests
+_set_ingress_environment_variables
+export INGRESS_URL="$INGRESS_HOST:$INGRESS_PORT"
+echo "$INGRESS_URL"
+for _ in {1..50}; do
+    curl -s "http://$INGRESS_URL/productpage" > /dev/null
+done
+
+# Now check Prometheus dashboard for the metric. It should be present
+function urlencode() {
+    python3 -c "import urllib.parse; print(urllib.parse.quote('''$1'''))"
+}
+
+function query_prometheus() {
+    local prometheus_api_root="localhost:9090/api/v1"
+    local encoded_query=$(urlencode "$1")
+    curl -sg -m 3.0 "http://$prometheus_api_root/query?query=$encoded_query"
+}
+
+function query_total_requests() {
+    query_prometheus "$snip_querying_istio_metrics_4" | jq '.data.result[0].metric.__name__'
+}
+
+function query_requests_to_productpage() {
+    query_prometheus "$snip_querying_istio_metrics_5" | jq '.data.result[0].metric.destination_service'
+}
+
+function query_requests_to_reviews_v3() {
+    query_prometheus "$snip_querying_istio_metrics_6" | jq '.data.result[0].metric.destination_service,.data.result[0].metric.destination_version'
+}
+
+function query_rate_of_requests_to_productpage_5m() {
+    query_prometheus "$snip_querying_istio_metrics_7" | jq '.data.result[0].metric.destination_service'
+}
+
+# Prometheus living inside cluster is not accessible from outside.
+# So we need some sort of port forwarding mechanism
+istioctl dashboard prometheus &
+
+_verify_contains query_total_requests '"istio_requests_total"'
+_verify_contains query_requests_to_productpage '"productpage.default.svc.cluster.local"'
+_verify_contains query_requests_to_reviews_v3 '"reviews.default.svc.cluster.local"'
+_verify_contains query_requests_to_reviews_v3 '"v3"'
+_verify_contains query_rate_of_requests_to_productpage_5m '"productpage.default.svc.cluster.local"'
+pgrep istioctl | xargs kill
+
+# @cleanup
+set +e
+pgrep istioctl | xargs kill
```

Comment:
Doesn't `snip_cleanup_1` work? Would be better if the test could run the commands as documented, if possible.

Reply:
No. It didn't work. That was the first one I tried. Originally it was only in cleanup section. Tests didn't terminate. While looking in some other place, I found similar pattern in `tcp-metrics` test and used it. Digging into commit https://github.com/istio/istio.io/pull/7826/files/63a826b94cbff6eec4975432e6f13faef81789e8#diff-fb7f37aa500c501bd8b4ec45f80023f2 it seems that it was an issue there as well. How about changing the docs (as this does the same thing) to match snips?

Reply:
The cleanup scripts fail because we don't have `killall` installed in the test image. We could add it, and then use the cleanup script.

What we do have in the install image is psgrep, so in my tests I resorted to using that. We could change the docs to use that and then it would be in the script and we just call the script.

However, and I haven't figured out why, we need to make the call twice else the container will fail to stop. It's something weird with our image since I can run everything from the command line and it works just fine. I suspect it's something weird in that we are running a script and that's starting the istioctl...

Reply:
To be honest, when I tried running commands one by one, it worked fine. But not while running automated tests. (Timing issue probably?)

Reply:
If the code works that was in the other test such that we can hit the prometheus server to verify metrics from another container (maybe we need to have a sleep container installed), so we don't use portforwarding and have to try and kill that process, that might be better. However, that makes the doc tougher since we would have to have them not use the dashboard command (or divert from the code which @frankbu rightfullysuggests is not good).

Reply:
Final answer for me would be to update the docs to use psgrep instead of killall and then revert to calling the script function twice.

Reply:
Correct. A single psgrep works fine if you are running it from the command line, but fails in the test case (either in the pipeline or local). Someone who understands the subprocesses and running in a script within a container may be able to explain why we need the second invocation in the test.

Reply:
> Final answer for me would be to update the docs to use psgrep instead of killall and then revert to calling the script function twice.

Cool. I'll update that.

Reply:
I thought I tried it, but we might also try doing a single call and add a `-9` to the kill in case after the first call it's trying to do a nice shutdown and the second is causing that to end anyway.

Reply:
> Someone who understands the subprocesses and running in a script within a container may be able to explain why we need the second invocation in the test.
Wait... But both are running in the same container (means same PID namespace). Right? On a different note, I ran into this - https://www.kernel.org/doc/Documentation/cgroup-v1/pids.txt . 

> The process number controller is used to allow a cgroup hierarchy to stop any
new tasks from being fork()'d or clone()'d after a certain limit is reached.
Do we have any such cgroup settings?

Reply:
Because of the weird issue of needing to call it twice in the test, I wouldn't update the doc yet. If it's working now, I think we should merge with the different test cleanup vs doc cleanup. We can revisit the doc cleanup instructions in a followup PR.

Reply:
Sure. That followup PR could clean up both tasks..